src/main/java/org/folio/sidecar/integration/keycloak/IntrospectionService.java

```diff
@@ -0,0 +1,114 @@
+import lombok.RequiredArgsConstructor;
+import lombok.extern.log4j.Log4j2;
+import org.folio.sidecar.integration.kafka.LogoutEvent;
+import org.folio.sidecar.integration.keycloak.configuration.KeycloakProperties;
+import org.folio.sidecar.integration.keycloak.model.TokenIntrospectionResponse;
+import org.folio.sidecar.model.ClientCredentials;
+import org.folio.sidecar.service.CacheInvalidatable;
+import org.folio.sidecar.service.store.AsyncSecureStore;
+
+@Log4j2
+@ApplicationScoped
+@RequiredArgsConstructor
+public class IntrospectionService implements CacheInvalidatable {
+
+  private final KeycloakProperties properties;
+  private final KeycloakClient keycloakClient;
+  private final AsyncSecureStore secureStore;
+  private final Cache<String, TokenIntrospectionResponse> tokenCache;
+
+  public Future<RoutingContext> checkActiveToken(RoutingContext ctx) {
+    return introspectToken(ctx)
+      .map(TokenIntrospectionResponse::isActive)
+      .flatMap(IntrospectionService::handleTokenStatus);
+  }
+
+  @Override
+  public void invalidate(LogoutEvent event) {
+    if (LOGOUT_ALL == event.getType()) {
+      tokenCache.asMap().forEach((key, value) -> {
+        if (key.contains(event.getUserId())) {
+          tokenCache.put(key, INACTIVE_TOKEN);
+        }
+      });
+      return;
+    }
+
+    tokenCache.asMap().forEach((key, value) -> {
+      if (key.contains(event.getUserId()) && key.contains(event.getSessionId())) {
+        tokenCache.put(key, INACTIVE_TOKEN);
+      }
+    });
+  }
+
+  private Future<TokenIntrospectionResponse> introspectToken(RoutingContext ctx) {
+    var parsedToken = getParsedToken(ctx);
+    if (parsedToken.isEmpty()) {
+      return failedFuture(new UnauthorizedException("Parsed token not found in request"));
+    }
+    var userId = getUserIdClaim(parsedToken.get());
+    if (userId.isEmpty()) {
+      return failedFuture(new UnauthorizedException("user_id claim not found in token"));
+    }
+
+    var tenant = getOriginTenant(ctx);
+    var sessionId = getSessionIdClaim(parsedToken.get());
+    var key = cacheKey(tenant, userId.get(), sessionId);
+
+    var cachedIntrospection = tokenCache.getIfPresent(key);
+    if (cachedIntrospection != null) {
+      return succeededFuture(cachedIntrospection);
+    }
+
+    var token = ctx.request().getHeader(TOKEN);
+    var clientId = tenant + properties.getLoginClientSuffix();
+    return secureStore.get(tenantStoreKey(tenant, clientId))
+      .map(clientSecret -> ClientCredentials.of(clientId, clientSecret))
+      .flatMap(client -> keycloakClient.introspectToken(tenant, client, token))
+      .flatMap(response -> handelAndCacheResponse(response, key));
+  }
+
+  private Future<TokenIntrospectionResponse> handelAndCacheResponse(HttpResponse<Buffer> response, String key) {
+    var statusCode = response.statusCode();
+    if (statusCode != 200) {
+      log.warn("Failed to introspect user token: response = {}", response::bodyAsString);
```

Comment:
It would make sense to also log HTTP status - body may be empty in case Keycloak returns 401/403 etc.